### PR TITLE
Map Merchant To Contact an Group CCC expense by expense id

### PR DESCRIPTION
**Pull request by NileshPant1999**
If there is no matching contact in xero we are mapping them to xero as 'Credit Card Misc'
https://go.xero.com/Bank/ViewTransaction.aspx?bankTransactionID=7077a0dc-41cc-4322-8132-67fc78499300&accountID=562555f2-8cde-4ce9-8203-0363922537a4

If there is a matching contact in xero for the given merchant we will be mapping contact corresponding to that merchant
https://go.xero.com/Bank/ViewTransaction.aspx?bankTransactionID=306a24b4-946f-42b4-8c3a-5fcc0aeeee47&accountID=562555f2-8cde-4ce9-8203-0363922537a4

The flow for reimbursable expenses is usual as before 
https://go.xero.com/AccountsPayable/View.aspx?invoiceID=7b4aa619-f336-43b2-9162-2f111e00a828

### CODE REVIEW

File: apps/xero/models.py [hunk 1]
```diff
@@ -231,12 +231,29 @@ def create_bank_transaction(expense_group: ExpenseGroup):
...
-            workspace_id=expense_group.workspace_id
-        ).destination.destination_id
+        if map_merchant_to_contact:
+            merchant = expense.vendor if expense.vendor else ''
+            
+            contact_id = DestinationAttribute.objects.filter(
+                value__iexact=merchant, attribute_type='CONTACT', workspace_id=expense_group.workspace_id
+            ).first()
+
+            expense_group.description['spent_at'] = expense.spent_at.strftime("%d-%m-%Y")
```
Comment (NileshPant1999) [inline apps/xero/models.py:241]: need confirmation on date format

File: apps/xero/tasks.py [hunk 2]
```diff
@@ -24,6 +24,31 @@
 logger = logging.getLogger(__name__)
 
 
+def get_or_create_credit_card_contact(workspace_id: int, merchant: str):
+    """
+    Get or create car default vendor
```
Comment (NileshPant1999) [inline apps/xero/tasks.py:29]: contact

File: apps/workspaces/utils.py [hunk 3]
```diff
@@ -79,10 +88,21 @@ def create_or_update_general_settings(general_settings_payload: Dict, workspace_
...
         }
     )
 
+    if general_settings.map_merchant_to_contact and \
+            general_settings.corporate_credit_card_expenses_object == 'BANK TRANSACTION':
+        expense_group_settings = ExpenseGroupSettings.objects.get(workspace_id=workspace_id)
+
+        ccc_expense_group_field = expense_group_settings.corporate_credit_card_expense_group_fields
+        ccc_expense_group_field.append('expense_id')
+        expense_group_settings.corporate_credit_card_expense_group_fields = ccc_expense_group_field
```
Comment (Shwetabhk) [inline apps/workspaces/utils.py:102]:
```suggestion
        expense_group_settings.corporate_credit_card_expense_group_fields = list(set(ccc_expense_group_field))
```

File: apps/workspaces/utils.py [hunk 4]
```diff
@@ -79,10 +88,21 @@ def create_or_update_general_settings(general_settings_payload: Dict, workspace_
...
+            'auto_create_destination_entity': general_settings_payload['auto_create_destination_entity'],
+            'map_merchant_to_contact': map_merchant_to_contact
         }
     )
 
+    if general_settings.map_merchant_to_contact and \
+            general_settings.corporate_credit_card_expenses_object == 'BANK TRANSACTION':
+        expense_group_settings = ExpenseGroupSettings.objects.get(workspace_id=workspace_id)
+
+        ccc_expense_group_field = expense_group_settings.corporate_credit_card_expense_group_fields
```
Comment (Shwetabhk) [inline apps/workspaces/utils.py:100]: Change name go `ccc_expense_group_fields`

File: apps/workspaces/migrations/0009_workspacegeneralsettings_map_merchant_to_contact.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('workspaces', '0008_workspacegeneralsettings_auto_create_destination_entity'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='workspacegeneralsettings',
+            name='map_merchant_to_contact',
```
Comment (ashwin1111) [inline apps/workspaces/migrations/0009_workspacegeneralsettings_map_merchant_to_contact.py:15]:
```suggestion
            name='testing1234',
```

File: apps/xero/utils.py [hunk 6]
```diff
@@ -37,6 +37,32 @@ def __init__(self, credentials_object: XeroCredentials, workspace_id: int):
...
+        Call xero api to get or create contact
+        :param email: email for contact user
+        :param contact_name: Name of the contact
+        :param create: False to just Get and True to Get or Create if not exists
+        :return: Contact
+        """
+        tenant_mapping = TenantMapping.objects.get(workspace_id=self.workspace_id)
+        self.connection.set_tenant_id(tenant_mapping.tenant_id)
+
+        contact_name = contact_name.replace('#', '%23')  # Replace '#' with %23
```
Comment (Shwetabhk) [inline apps/xero/utils.py:52]: Do we not need to handle single quote `'`?

Comment (NileshPant1999) [reply apps/xero/utils.py:52]: no need to handle single quotes working fine for xero

`connection.contacts.search_contact_by_contact_name('Nilesh"s Company')`

`{'ContactID': 'bce8795f-4859-4f9c-8d9a-54f5b875d7c0', 'ContactStatus': 'ACTIVE', 'Name': "Nilesh's Company", 'EmailAddress': '', 'BankAccountDetails': '', 'Addresses': [{'AddressType': 'STREET', 'City': '', 'Region': '', 'PostalCode': '', 'Country': ''}, {'AddressType': 'POBOX', 'City': '', 'Region': '', 'PostalCode': '', 'Country': ''}], 'Phones': [{'PhoneType': 'DDI', 'PhoneNumber': '', 'PhoneAreaCode': '', 'PhoneCountryCode': ''}, {'PhoneType': 'DEFAULT', 'PhoneNumber': '', 'PhoneAreaCode': '', 'PhoneCountryCode': ''}, {'PhoneType': 'FAX', 'PhoneNumber': '', 'PhoneAreaCode': '', 'PhoneCountryCode': ''}, {'PhoneType': 'MOBILE', 'PhoneNumber': '', 'PhoneAreaCode': '', 'PhoneCountryCode': ''}], 'UpdatedDateUTC': '/Date(1617862083580+0000)/', 'ContactGroups': [], 'IsSupplier': False, 'IsCustomer': False, 'ContactPersons': [], 'HasAttachments': False, 'HasValidationErrors': False}`

Comment (NileshPant1999) [reply apps/xero/utils.py:52]: need working in double quotes getting some issue there i will try to fix that

File: apps/xero/utils.py [hunk 7]
```diff
@@ -37,6 +37,33 @@ def __init__(self, credentials_object: XeroCredentials, workspace_id: int):
...
+        contact_name = contact_name.replace('"', '') #replace '"' with ''
```
Comment (ashwin1111) [inline apps/xero/utils.py:53]:
```suggestion
        contact_name = contact_name.replace('"', '') # remove double quotes from merchant name
```

**Review by ashwin1111 — APPROVED** [review]

**Review by Sravanksk — APPROVED** [review]

**Pull request merged**